src/org/apache/xpath/functions/FuncForEach.java

```diff
@@ -0,0 +1,212 @@
+import org.w3c.dom.Node;
+
+/**
+ * Execute the for-each() function.
+ * 
+ * @author Mukul Gandhi <mukulg@apache.org>
+ * 
+ * @xsl.usage advanced
+ */
+/*
+ * fn:for-each is one of XPath 3.1's higher-order function
+ * (ref, https://www.w3.org/TR/xpath-functions-31/#higher-order-functions).
+ * 
+ * The XPath function fn:for-each has following signature, and definition,
+ * 
+ * fn:for-each($seq as item()*, $action as function(item()) as item()*) as item()*
+ * 
+ * The function fn:for-each applies the function item $action to every item from 
+ * the sequence $seq in turn, returning the concatenation of the resulting sequences
+ * in order.
+ */
+public class FuncForEach extends Function2Args {
+
+   private static final long serialVersionUID = 2912594883291006421L;
+   
+   private static final String FUNCTION_NAME = "for-each()"; 
+
+  /**
+   * Execute the function. The function must return a valid object.
+   * 
+   * @param xctxt The current execution context.
+   * 
+   * @return A valid XObject.
+   *
+   * @throws javax.xml.transform.TransformerException
+   */
+  public XObject execute(XPathContext xctxt) throws javax.xml.transform.TransformerException
+  {                      
+      
+        XObject funcEvaluationResult = null;
+        
+        final int contextNode = xctxt.getCurrentNode();
+        
+        Expression arg0 = getArg0();
+        Expression arg1 = getArg1();                
+        
+        DTMIterator arg0DtmIterator = null;
+        
+        XObject arg0XsObject = null;
+                  
+        if (arg0 instanceof LocPathIterator) {
+            arg0DtmIterator = arg0.asIterator(xctxt, contextNode);               
+        }
+        else {
+            arg0XsObject = arg0.execute(xctxt, contextNode);
+        }
+        
+        ResultSequence resultSeq = new ResultSequence();
+                    
+        if (arg1 instanceof InlineFunction) {            
+            resultSeq = evaluateFnForEach(xctxt, arg0XsObject, arg0DtmIterator, (InlineFunction)arg1); 
+        }
+        else if (arg1 instanceof Variable) {
+            XObject arg1VarValue = arg1.execute(xctxt);
+            if (arg1VarValue instanceof InlineFunction) {
+                resultSeq = evaluateFnForEach(xctxt, arg0XsObject, arg0DtmIterator, (InlineFunction)arg1VarValue);   
+            }
+            else {
+                throw new javax.xml.transform.TransformerException("FORG0006 : The second argument to function call for-each(), "
+                                                                                               + "is not a function item.", xctxt.getSAXLocator());    
+            }
+        }
+        else {
+            throw new javax.xml.transform.TransformerException("FORG0006 : The second argument to function call for-each(), "
+                                                                                           + "is not a function item.", xctxt.getSAXLocator());               
```

Comment:
The errors have the same text yet they are from different branches. It would be hard to tell what went wrong